### PR TITLE
Upgrade containerapps SDK to 2023-04-01-preview

### PR DESCRIPTION
**Pull request by Juliehzl**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ContainerAppJobExecutionBase.cs [hunk 1]
```diff
@@ -11,7 +11,7 @@ namespace Azure.ResourceManager.AppContainers.Models
     public partial class ContainerAppJobExecutionBase
     {
         /// <summary> Initializes a new instance of ContainerAppJobExecutionBase. </summary>
-        public ContainerAppJobExecutionBase()
+        internal ContainerAppJobExecutionBase()
```
Comment (archerzz) [inline sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ContainerAppJobExecutionBase.cs:14]: Breaking change. Looks like this model is changed to abstract?

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ContainerAppJobExecutionBase.cs [hunk 2]
```diff
@@ -25,8 +25,8 @@ internal ContainerAppJobExecutionBase(string name, string id)
         }
 
         /// <summary> Job execution name. </summary>
-        public string Name { get; set; }
```
Comment (archerzz) [inline sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ContainerAppJobExecutionBase.cs:28]: breaking change. setter is removed

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ContainerAppJobTriggerType.cs [hunk 3]
```diff
@@ -22,12 +22,12 @@ public ContainerAppJobTriggerType(string value)
             _value = value ?? throw new ArgumentNullException(nameof(value));
         }
 
-        private const string ScheduledValue = "Scheduled";
+        private const string ScheduleValue = "Schedule";
         private const string EventValue = "Event";
         private const string ManualValue = "Manual";
 
-        /// <summary> Scheduled. </summary>
-        public static ContainerAppJobTriggerType Scheduled { get; } = new ContainerAppJobTriggerType(ScheduledValue);
```
Comment (archerzz) [inline sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ContainerAppJobTriggerType.cs:30]: breaking change, an enum value is renamed

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/JobExecutionNamesCollection.cs [hunk 4]
```diff
@@ -1,31 +0,0 @@
...
-// <auto-generated/>
-
-#nullable disable
-
-using System;
-using System.Collections.Generic;
-using System.Linq;
-using Azure.Core;
-
-namespace Azure.ResourceManager.AppContainers.Models
```
Comment (archerzz) [inline sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/JobExecutionNamesCollection.cs:13]: breaking change, a model is removed.

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/JobScale.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        internal JobScale(int? pollingInterval, int? minExecutions, int? maxExecutions, IList<JobScaleRule> rules)
+        {
+            PollingInterval = pollingInterval;
+            MinExecutions = minExecutions;
+            MaxExecutions = maxExecutions;
+            Rules = rules;
+        }
+
+        /// <summary> Interval to check each event source in seconds. Defaults to 30s. </summary>
+        public int? PollingInterval { get; set; }
```
Comment (archerzz) [inline sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/JobScale.cs:36]: What's the unit? `InSeconds` or `InMinutes` or else? Better to append the unit to the name.

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ServiceBind.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        /// <param name="serviceId"> Resource id of the target service. </param>
+        /// <param name="name"> Name of the service bind. </param>
+        internal ServiceBind(string serviceId, string name)
+        {
+            ServiceId = serviceId;
+            Name = name;
+        }
+
+        /// <summary> Resource id of the target service. </summary>
+        public string ServiceId { get; set; }
```
Comment (archerzz) [inline sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ServiceBind.cs:28]: What's the format of this property value?
- If it's a GUID, use type [Guid](https://learn.microsoft.com/en-us/dotnet/api/system.guid?view=net-6.0). See example [here](https://apiview.dev/Assemblies/Review/25a98e223bda4f39b4ef76bebffa2c23#d4cebe383870410cab2e14d0709f78bb).
- If it's a resource identifier, use type [ResourceIdentifier](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/ResourceIdentifier.cs). See example [here](https://apiview.dev/Assemblies/Review/25a98e223bda4f39b4ef76bebffa2c23#9f6fc0ae7c8e497e98d7979dadf97a4c).

**Review by archerzz — COMMENTED** [review]
There are quite many breaking changes. Some of them can hardly be worked around. @ArthurMa1978 Do you think we should bump the major version of `containerapps` SDK?

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ServiceBind.cs (hunk 6 above)
Comment (Juliehzl) [reply sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ServiceBind.cs:28]: As mentioned in summary, it is resource id

**Comment by ArthurMa1978** [comment]
> There are quite many breaking changes. Some of them can hardly be worked around. @ArthurMa1978 Arthur Ma FTE Do you think we should bump the major version of `containerapps` SDK?

No, all breaking changes need be fixed by customization code

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/ContainerAppJobExecutionData.cs [hunk 7]
```diff
@@ -7,28 +7,32 @@
...
-    public partial class ContainerAppJobExecution : ResourceData
+    /// <summary>
+    /// A class representing the ContainerAppJobExecution data model.
+    /// Container Apps Job execution.
+    /// </summary>
+    public partial class ContainerAppJobExecutionData : ResourceData
     {
-        /// <summary> Initializes a new instance of ContainerAppJobExecution. </summary>
-        internal ContainerAppJobExecution()
+        /// <summary> Initializes a new instance of ContainerAppJobExecutionData. </summary>
```
[the remark this answers is not in the document]
Comment (vinisoto) [reply sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/ContainerAppJobExecutionData.cs:21]: @archerzz The name `ContainerAppJobExecutionData` doesn't come from our swagger. Do we know where is it coming from? Our previous SDK version had `ContainerAppJobExecution` so this must be a new thing maybe?

Should we just manually fix it? (will this error happen again in the next release?)

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/ContainerAppJobResource.cs [hunk 8]
```diff
@@ -377,84 +420,6 @@ public virtual ArmOperation<ContainerAppJobExecutionBase> Start(WaitUntil waitUn
...
-        /// </item>
-        /// </list>
-        /// </summary>
-        /// <param name="waitUntil"> <see cref="WaitUntil.Completed"/> if the method should wait to return until the long-running operation has completed on the service; <see cref="WaitUntil.Started"/> if it should return after starting the operation. For more information on long-running operations, please see <see href="https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/LongRunningOperations.md"> Azure.Core Long-Running Operation samples</see>. </param>
-        /// <param name="jobExecutionName"> Job execution name. </param>
-        /// <param name="cancellationToken"> The cancellation token to use. </param>
-        /// <exception cref="ArgumentException"> <paramref name="jobExecutionName"/> is an empty string, and was expected to be non-empty. </exception>
-        /// <exception cref="ArgumentNullException"> <paramref name="jobExecutionName"/> is null. </exception>
-        public virtual async Task<ArmOperation> StopExecutionAsync(WaitUntil waitUntil, string jobExecutionName, CancellationToken cancellationToken = default)
```
[the remark this answers is not in the document]
Comment (jijohn14) [reply sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/ContainerAppJobResource.cs:398]: Auto rest now creates a new `ContainerAppJobExecutionResource` to handle all operations related to it. This is the correct way which is fixed in this api-version

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/ContainerAppJobExecutionData.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jijohn14) [reply sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/ContainerAppJobExecutionData.cs:21]: Auto rest now creates a new `ContainerAppJobExecutionResource` to handle all operations related to it, which requires `ContainerAppJobExecutionData`. This is the correct way which is fixed in this api-version

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/JobScale.cs (hunk 5 above)
Comment (jijohn14) [reply sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/JobScale.cs:36]: Agreed. Can we make this change without breaking swagger spec? We don't want to manually make the change for every.net sdk generation.

**Comment by jijohn14** [comment]
> > There are quite many breaking changes. Some of them can hardly be worked around. @ArthurMa1978 Arthur Ma FTE Do you think we should bump the major version of `containerapps` SDK?
> 
> No, all breaking changes need be fixed by customization code

All breaking changes are intentional fixes. Note that all versions are preview and will be deprecated end of this year.
We want to have the correct pattern for next stable version release.

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ContainerAppJobTriggerType.cs (hunk 3 above)
Comment (anandanthony) [reply sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ContainerAppJobTriggerType.cs:30]: The enum was changed to have the correct value, we discussed and got this breaking change approved:
https://github.com/Azure/azure-rest-api-specs/pull/24400

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ContainerAppJobExecutionBase.cs (hunk 1 above)
Comment (anandanthony) [reply sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ContainerAppJobExecutionBase.cs:14]: This was not used - `JobExecutionNamesCollection` and was removed.

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/JobExecutionNamesCollection.cs (hunk 4 above)
Comment (anandanthony) [reply sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/JobExecutionNamesCollection.cs:13]: This was not used - `JobExecutionNamesCollection`  and was removed.

File: sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ContainerAppJobExecutionBase.cs (hunk 1 above)
Comment (anandanthony) [reply sdk/containerapps/Azure.ResourceManager.AppContainers/src/Generated/Models/ContainerAppJobExecutionBase.cs:14]: `internal` is the correct modifier here.

**Review by archerzz — APPROVED** [review]

**Pull request merged**